File: app/src/main/java/edu/aku/hassannaqvi/mi_covid/ui/sections/SectionDActivity.java
package edu.aku.hassannaqvi.mi_covid.ui.sections;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.validatorcrawler.aliazaz.Clear;
import com.validatorcrawler.aliazaz.Validator;

import org.json.JSONException;
import org.json.JSONObject;

import edu.aku.hassannaqvi.mi_covid.R;
import edu.aku.hassannaqvi.mi_covid.contracts.FormsContract;
import edu.aku.hassannaqvi.mi_covid.core.DatabaseHelper;
import edu.aku.hassannaqvi.mi_covid.core.MainApp;
import edu.aku.hassannaqvi.mi_covid.databinding.ActivitySectionDBinding;
import edu.aku.hassannaqvi.mi_covid.ui.other.EndingActivity;

public class SectionDActivity extends AppCompatActivity {

    ActivitySectionDBinding bi;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_section_d);
        setupSkips();
    }


    private void setupSkips() {

        bi.d02.setOnCheckedChangeListener(((radioGroup, i) -> {
            if (i == bi.d0202.getId()) {
                Clear.clearAllFields(bi.fldGrpCVd03);
            }
        }));

        /*bi.ah2.setOnCheckedChangeListener(((radioGroup, i) -> {
            if (i == bi.ah2b.getId()) {
                Clear.clearAllFields(bi.fldGrpCVah3);
            }
        }));*/

    }


    private boolean UpdateDB() {
        DatabaseHelper db = MainApp.appInfo.getDbHelper();
        int updcount = db.updatesFormColumn(FormsContract.FormsTable.COLUMN_SD, MainApp.form.getsD());
        if (updcount > 0) {
            return true;
        } else {
            Toast.makeText(this, "Updating Database... ERROR!", Toast.LENGTH_SHORT).show();
            return false;
        }
    }


    public void BtnContinue() {
        if (formValidation()) {
            try {
                SaveDraft();
            } catch (JSONException e) {
                e.printStackTrace();
            }
            if (UpdateDB()) {
                finish();
                startActivity(new Intent(this, SectionEActivity.class));
            } else {
                Toast.makeText(this, "Failed to Update Database!", Toast.LENGTH_SHORT).show();
            }
        }
    }


    private void SaveDraft() throws JSONException {


        JSONObject json = new JSONObject();
        json.put("d1", bi.d1.getText().toString());

        json.put("d02", bi.d0201.isChecked() ? "1"
                : bi.d0202.isChecked() ? "2"
                :  "-1");

        json.put("d03", bi.d0301.isChecked() ? "1"
                : bi.d0302.isChecked() ? "2"
                : bi.d0303.isChecked() ? "3"
                : bi.d0304.isChecked() ? "4"
                : bi.d0305.isChecked() ? "5"
                : bi.d03096.isChecked() ? "96"
                :  "-1");

        json.put("d04", bi.d0401.isChecked() ? "1"
                : bi.d0402.isChecked() ? "2"
                :  "-1");

        json.put("d05", bi.d0501.isChecked() ? "1"
                : bi.d0502.isChecked() ? "2"
                : "-1");


    }


    private boolean formValidation() {
        return Validator.emptyCheckingContainer(this, bi.fldGrpSectionD);
    }


    public void BtnEnd() {
        startActivity(new Intent(this, EndingActivity.class).putExtra("complete", false));
    }


    @Override
    public void onBackPressed() {
        Toast.makeText(this, "You can't go back", Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/edu/aku/hassannaqvi/mi_covid/ui/sections/SectionDActivity.kt
package edu.aku.hassannaqvi.mi_covid.ui.sections

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.RadioGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.validatorcrawler.aliazaz.Clear
import com.validatorcrawler.aliazaz.Validator.Companion.emptyCheckingContainer
import edu.aku.hassannaqvi.mi_covid.CONSTANTS
import edu.aku.hassannaqvi.mi_covid.R
import edu.aku.hassannaqvi.mi_covid.contracts.FormsContract
import edu.aku.hassannaqvi.mi_covid.core.MainApp.appInfo
import edu.aku.hassannaqvi.mi_covid.core.MainApp.form
import edu.aku.hassannaqvi.mi_covid.databinding.ActivitySectionDBinding
import edu.aku.hassannaqvi.mi_covid.utils.contextBackActivity
import edu.aku.hassannaqvi.mi_covid.utils.openFormEndActivity
import edu.aku.hassannaqvi.mi_covid.utils.showTooltip
import org.json.JSONException

class SectionDActivity : AppCompatActivity() {
    lateinit var bi: ActivitySectionDBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bi = DataBindingUtil.setContentView(this, R.layout.activity_section_d)
        bi.callback = this
        bi.form = form
        setupSkips()
    }

    private fun setupSkips() {
        bi.d02.setOnCheckedChangeListener { radioGroup: RadioGroup, i: Int ->
            if (i == bi.d0202.id) {
                Clear.clearAllFields(bi.fldGrpCVd03)
            }
        }
    }

    @Throws(JSONException::class)
    private fun saveDraft() {

        form.d01 = bi.d01.text.toString()
        form.d02 = when {
            bi.d0201.isChecked -> "1"
            bi.d0202.isChecked -> "2"
            else -> "-1"
        }
        form.d03 = when {
            bi.d0301.isChecked -> "1"
            bi.d0302.isChecked -> "2"
            bi.d0303.isChecked -> "3"
            bi.d0304.isChecked -> "4"
            bi.d0305.isChecked -> "5"
            bi.d03096.isChecked -> "96"
            else -> "-1"
        }
        form.d04 = when {
            bi.d0401.isChecked -> "1"
            bi.d0402.isChecked -> "2"
            else -> "-1"
        }
        form.d05 = when {
            bi.d0501.isChecked -> "1"
            bi.d0502.isChecked -> "2"
            else -> "-1"
        }

        /*  val json = JSONObject()
          json.put("d1", bi.d1.text.toString())
          json.put("d02", when {
              bi.d0201.isChecked -> "1"
              bi.d0202.isChecked -> "2"
              else -> "-1"
          })
          json.put("d03", when {
              bi.d0301.isChecked -> "1"
              bi.d0302.isChecked -> "2"
              bi.d0303.isChecked -> "3"
              bi.d0304.isChecked -> "4"
              bi.d0305.isChecked -> "5"
              bi.d03096.isChecked -> "96"
              else -> "-1"
          })
          json.put("d04", when {
              bi.d0401.isChecked -> "1"
              bi.d0402.isChecked -> "2"
              else -> "-1"
          })
          json.put("d05", when {
              bi.d0501.isChecked -> "1"
              bi.d0502.isChecked -> "2"
              else -> "-1"
          })

          MainApp.form.setsD(json.toString())*/
    }


    private fun formValidation(): Boolean {
        return emptyCheckingContainer(this, bi.fldGrpSectionD)
    }


    private fun updateDB(): Boolean {
        val db = appInfo.dbHelper
        val updcount = db.updatesFormColumn(FormsContract.FormsTable.COLUMN_SD, form.sBtoString())
        return if (updcount > 0) {
            true
        } else {
            Toast.makeText(this, "Sorry. You can't go further.\n Please contact IT Team (Failed to update DB)", Toast.LENGTH_SHORT).show()
            false
        }
    }


    fun BtnContinue(v: View) {
        if (!formValidation()) return
        try {
            saveDraft()
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        if (updateDB()) {
            finish()
            startActivity(Intent(this, SectionEActivity::class.java))
        } else {
            Toast.makeText(this, "Sorry. You can't go further.\n Please contact IT Team (Failed to update DB)", Toast.LENGTH_SHORT).show()
        }
    }


    fun BtnEnd(v: View) {
        openFormEndActivity(this, CONSTANTS.FSTATUS_END_FLAG, 2)
    }


    override fun onBackPressed() {
        contextBackActivity(this)
    }

    fun showTooltipView(view: View?) {
        showTooltip(this, view!!)
    }
}